### PR TITLE
feat(balancer): add latency balancer algorithm

### PR DESCRIPTION
**Pull request by attenuation**
### Summary

add latency balancer algorithm

FTI-4099
KAG-79

### Full changelog
1. change the balancer system function code to support updating the balancer status in the log phase
2. introduce the latency balancer algorithm

https://linkerd.io/2016/03/16/beyond-round-robin-load-balancing-for-latency/

### CODE REVIEW

**Comment by chronolaw** [comment]
Needs a change log entry.

File: kong/runloop/balancer/ewma.lua
Comment (chronolaw) [inline kong/runloop/balancer/ewma.lua:28]: localize `math.exp`

style:
```
math_exp(-td / DECAY_TIME)
```

File: kong/runloop/balancer/ewma.lua
Comment (chronolaw) [inline kong/runloop/balancer/ewma.lua:40]: could be `local address_count = 0`, and set to `self` at the end.

File: kong/runloop/balancer/ewma.lua
Comment (chronolaw) [inline kong/runloop/balancer/ewma.lua:109]: localize `ngx.var`

File: kong/runloop/balancer/ewma.lua
Comment (chronolaw) [inline kong/runloop/balancer/ewma.lua:106]: should use `tonumber`

File: kong/runloop/balancer/ewma.lua
Comment (chronolaw) [inline kong/runloop/balancer/ewma.lua:62]: we can use `table.clear` to reuse this table.

File: kong/runloop/balancer/ewma.lua
Comment (chronolaw) [inline kong/runloop/balancer/ewma.lua:179]: add one line: `local address_count = self.address_count`

File: kong/runloop/balancer/ewma.lua
Comment (chronolaw) [inline kong/runloop/balancer/ewma.lua:193]: should use `table.nkeys` and cache it.

File: kong/runloop/balancer/ewma.lua
Comment (chronolaw) [inline kong/runloop/balancer/ewma.lua:194]: localize `ngx.WARN`

File: kong/runloop/balancer/ewma.lua
Comment (chronolaw) [inline kong/runloop/balancer/ewma.lua:147]: could we change these params `cacheOnly` `valueToHash` to snake_case?

File: kong/runloop/balancer/ewma.lua
Comment (chronolaw) [inline kong/runloop/balancer/ewma.lua:158]: add a `,` at the end of line.

File: kong/db/schema/entities/upstreams.lua
Comment (chronolaw) [inline kong/db/schema/entities/upstreams.lua:329]: Why remove this line?

File: kong/runloop/balancer/ewma.lua
Comment (chronolaw) [inline kong/runloop/balancer/ewma.lua:80]: It could be 
```
if slow_start_ewma == nil then
  return
end
```

**Comment by chronolaw** [comment]
I think the code needs more clean work to do, let it fit kong's code style.

File: kong/runloop/balancer/ewma.lua
Comment (chronolaw) [inline kong/runloop/balancer/ewma.lua:231]: blank line before `else`

File: kong/runloop/balancer/ewma.lua
Comment (chronolaw) [inline kong/runloop/balancer/ewma.lua:184]: `tonumber` seems unnecessary.

File: kong/runloop/balancer/ewma.lua
Comment (colinh-kong) [inline kong/runloop/balancer/ewma.lua:20]: should these be configurable with sensible defaults?

File: kong/runloop/balancer/ewma.lua
Comment (colinh-kong) [inline kong/runloop/balancer/ewma.lua:107]: same as above should use `tonumber`. 

Just curious would it ever be possible for `upstream_response_time` or `upstream_connect_time` to not be set?

File: kong/runloop/balancer/ewma.lua
Comment (colinh-kong) [inline kong/runloop/balancer/ewma.lua:91]: is using `or` like this safe / suitable?

File: spec/02-integration/05-proxy/10-balancer/07-ewma_spec.lua
Comment (colinh-kong) [inline spec/02-integration/05-proxy/10-balancer/07-ewma_spec.lua:67]: why 100?

**Comment by colinh-kong** [comment]
Do we have an early documentation PR?

File: kong/runloop/balancer/ewma.lua
Comment (locao) [inline kong/runloop/balancer/ewma.lua:134]: Why are the addresses shuffled before getting the lowest score?

File: kong/runloop/balancer/ewma.lua
Comment (locao) [inline kong/runloop/balancer/ewma.lua:95]: Style: if we change this to `if update then` we have a single exit-point function. Not required, but easier to read.

File: kong/runloop/balancer/ewma.lua
Comment (locao) [reply kong/runloop/balancer/ewma.lua:231]: The `else` statement could be removed at all.

File: kong/runloop/balancer/ewma.lua
Comment (locao) [inline kong/runloop/balancer/ewma.lua:108]: Could this be changed to `ctx.KONG_PROXY_LATENCY`? It's ready to be used (no need to deal with `tonumber` or `ngx.var`).

File: kong/runloop/balancer/ewma.lua
Comment (locao) [inline kong/runloop/balancer/ewma.lua:114]: What are we trying to avoid here? `upstream` is not used.

File: kong/runloop/balancer/init.lua
Comment (locao) [inline kong/runloop/balancer/init.lua:469]: Personally I don't like this function being called `log`, it seems ambiguous to what it is actually doing.

File: spec/01-unit/09-balancer/06-ewma_spec.lua
Comment (locao) [inline spec/01-unit/09-balancer/06-ewma_spec.lua:12]: Commented code
```suggestion
```

File: spec/01-unit/09-balancer/06-ewma_spec.lua
Comment (locao) [inline spec/01-unit/09-balancer/06-ewma_spec.lua:16]: Ditto
```suggestion
```

**Review by locao — CHANGES_REQUESTED** [review]

File: kong/runloop/balancer/ewma.lua
Comment (attenuation) [reply kong/runloop/balancer/ewma.lua:28]: @colinh-kong I think these config no need to configurable at this moment.

File: kong/runloop/balancer/ewma.lua
Comment (attenuation) [reply kong/runloop/balancer/ewma.lua]: new_address is mean new version address array compare to ewma context (maybe)

File: kong/runloop/balancer/ewma.lua
Comment (attenuation) [reply kong/runloop/balancer/ewma.lua:106]: if `self.ewma[address] = nil` it will set to 0

File: spec/02-integration/05-proxy/10-balancer/07-ewma_spec.lua
Comment (attenuation) [reply spec/02-integration/05-proxy/10-balancer/07-ewma_spec.lua:67]: It's just spec test.

File: kong/runloop/balancer/ewma.lua
Comment (attenuation) [reply kong/runloop/balancer/ewma.lua:149]: consider this situation, when the ewma balancer just starts, all ewma score is 0, we need to select one node randomly. not the first node.

File: kong/runloop/balancer/ewma.lua
Comment (attenuation) [reply kong/runloop/balancer/ewma.lua:123]: `ctx.KONG_PROXY_LATENCY` include kong code execute time, ewma just need `connect_time + response_time`
https://github.com/Kong/kong/blob/d7e3cbc7ac2f28fa0cc057859203b01e61f993f7/kong/init.lua#L1114

File: kong/runloop/balancer/ewma.lua
Comment (attenuation) [reply kong/runloop/balancer/ewma.lua:129]: yes, in Nginx https://github.com/nginx/nginx/blob/9c7a2c7ce4ad02a36df1bb0ee5e40a84610fffb9/src/http/ngx_http_upstream.c#L5605, if upstream not set, log phase always execute (no matter no upstream was set), if these situations happen, `upstream_addr` will be `nil`

**Comment by attenuation** [comment]
@chronolaw @locao Sorry guys, I forgot to check my code style

File: kong/runloop/balancer/ewma.lua
Comment (chronolaw) [inline kong/runloop/balancer/ewma.lua:31]: style: alignment.

File: kong/runloop/balancer/ewma.lua
Comment (chronolaw) [inline kong/runloop/balancer/ewma.lua:39]: style : `-td / DECAY_TIME)`

Comment (chronolaw) [reply kong/runloop/balancer/ewma.lua:39]: also, in kong, we use two space indent.

File: kong/runloop/balancer/ewma.lua
Comment (chronolaw) [inline kong/runloop/balancer/ewma.lua:62]: should be here: `self.address_count = address_count`

File: kong/runloop/balancer/ewma.lua
Comment (locao) [reply kong/runloop/balancer/ewma.lua:150]: Could you get the same result with something along the line of: `local lowest_score_index = math.random(1, k)`? We would avoid a `for` loop on every `getPeer()` call.

**Comment by attenuation** [comment]
holding if this pr is approved, I will change the ewma name to latency because the PM requirement

File: kong/runloop/balancer/ewma.lua
Comment (attenuation) [reply kong/runloop/balancer/ewma.lua]: No, it's not expected, I have fixed it.

**Review by locao — COMMENTED** [review]
This LGTM, let's just wait for the decision on the name before merging.

**Review by locao — APPROVED** [review]

**Pull request merged**
